### PR TITLE
C, C++, Python, C# API update for on device training

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request:

- Updates the existing `SaveCheckpoint` function signature.
   - `SaveCheckpoint` should now be called as a standalone function and does not belong to the training session anymore.
   - Example usage:

```cpp

Ort::CheckpointState checkpoint_state = Ort::CheckpointState::LoadCheckpoint(load_checkpoint_path);
// The checkpoint state includes the entire training session state and hence must outlive the training session 
Ort::TrainingSession training_session = Ort::TrainingSession(Ort::SessionOptions(), checkpoint_state, model_uri);
/* ... */
// Once training is done, the same checkpoint can be saved to a file
Ort::CheckpointState::SaveCheckpoint(checkpoint_state , save_checkpoint_path);

```

- Adds ability to add and get properties from the state.
   - Example usage:

```cpp
int64_t hours_in_a_year= 365 * 24;
state.AddProperty("hours in a year", hours_in_a_year);
auto property_i = state.GetProperty("hours in a year");

float pi = 3.14f;
state.AddProperty("pi", pi);
auto property_f = state.GetProperty("pi");

std::string best_framework("onnxruntime");
state.AddProperty("best framework", best_framework);
auto property_s = state.GetProperty("best framework");
```

- Updates the python bindings with the above two features.
   - `save_checkpoint` example usage:

```py
state = CheckpointState.load_checkpoint(checkpoint_file_path)
model = Module(training_model_file_path, state)
# ...
CheckpointState.save_checkpoint(session_state, checkpoint_save_path)

```
   - `add_property` and `get_property` example usage:

```py
state = CheckpointState.load_checkpoint(checkpoint_file_path)

state["property"] = property_value
assert "property" in state
assert state["property"] == property_value
```

- Adds detailed documentation for python bindings.
- Enables python bindings test in CI pipeline.
- Updates C# bindings for `SaveCheckpoint` method to reflect changes in C api. Missing C# bindings will be added in a follow up PR.

### CODE REVIEW

File: orttraining/orttraining/python/training/api/checkpoint_state.py [hunk 1]
```diff
@@ -1,19 +1,74 @@
...
+    def save_checkpoint(state: CheckpointState, checkpoint_uri: str | os.PathLike) -> None:
+        """Saves the checkpoint state to the checkpoint file
+
+        Args:
+            state: The checkpoint state object.
+            checkpoint_uri: The path to the checkpoint file.
         """
-        Initializes CheckpointState object with the given checkpoint uri.
-        The returned object will be used to initialize the Module.
+        C.save_checkpoint(state._state, str(checkpoint_uri))
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/api/checkpoint_state.py:43]: #15530 I just learn about this today

```suggestion
        C.save_checkpoint(state._state, os.fspath(checkpoint_uri))
```

File: orttraining/orttraining/python/training/api/checkpoint_state.py [hunk 2]
```diff
@@ -1,19 +1,74 @@
...
-    This class is a wrapper of CheckpointState Class.
-    """
 
-    def __init__(self, ckpt_uri) -> None:
+    def __init__(self, state: C.CheckpointState):
+        if not isinstance(state, C.CheckpointState):
+            raise TypeError(f"Invalid argument for CheckpointState received {type(state)}")
+        self._state = state
+
+    @staticmethod
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/api/checkpoint_state.py:23]: Consider a classmethod?

File: orttraining/orttraining/python/training/api/checkpoint_state.py [hunk 3]
```diff
@@ -1,19 +1,74 @@
...
+    def load_checkpoint(checkpoint_uri: str | os.PathLike) -> CheckpointState:
+        """Loads the checkpoint state from the checkpoint file
+
+        Args:
+            checkpoint_uri: The path to the checkpoint file.
+
+        Returns:
+            CheckpointState: The checkpoint state object.
+        """
+        return CheckpointState(C.load_checkpoint(str(checkpoint_uri)))
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/api/checkpoint_state.py:33]:
```suggestion
        return cls(C.load_checkpoint(os.fspath(checkpoint_uri)))
```

**Comment by AdamLouly** [comment]
Do we need to change examples here : https://github.com/microsoft/onnxruntime-training-examples
after the approval of this PR?

**Comment by baijumeswani** [comment]
> Do we need to change examples here : https://github.com/microsoft/onnxruntime-training-examples after the approval of this PR?

Yes, I will update the training examples once this PR is merged.

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 4]
```diff
@@ -37,6 +37,24 @@ public CheckpointState(string checkpointPath)
             }
         }
 
+        /// <summary>
+        /// Creates CheckpointState by loading from the handle
+        /// <param name="checkpointPath"> absolute path to checkpoint file.</param>
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:42]: nit: wrong comment

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs [hunk 5]
```diff
@@ -37,6 +37,24 @@ public CheckpointState(string checkpointPath)
...
+        /// </summary>
+        public CheckpointState(IntPtr checkpointHandle)
+            : base(checkpointHandle, true)
```
Comment (askhade) [inline csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs:45]: when do we need this api?

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs (hunk 4 above)
Comment (baijumeswani) [reply csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs]: This is no longer needed.

File: csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs (hunk 5 above)
Comment (baijumeswani) [reply csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs]: We needed it when `GetState` constructed a new CheckpointState object. We no longer need this.

**Review by askhade — APPROVED** [review]
LGTM!

A general note: We need to improve the coverage for c# tests. For example, add tests for RunEvalStep, Adding and getting properties etc...

**Comment by baijumeswani** [comment]
Sure. I intend to add more C# coverage in a follow up pull request. This one is mostly C, C++, Python.

**Comment by baijumeswani** [comment]
Thanks for the review 😄

**Comment by ahlzouao** [comment]
Hello @baijumeswani, 
I have been waiting for this feature on Onnxruntime since a long time, and that's a great news that you have enabled it :+1: Do you have any ideas when the https://github.com/microsoft/onnxruntime-training-examples will be updated with C and Python examples ? 
Thanks,

**Pull request merged**